mgard: Add configuration for compiling mgard with hip

kmorel opened the pull request:
<!--  
Remember that `spackbot` can help with your PR in multiple ways:
- `@spackbot help` shows all the commands that are currently available
- `@spackbot fix style` tries to push a commit to fix style issues in this PR
- `@spackbot re-run pipeline` runs the pipelines again, if you have write access to the repository 
-->

kmorel committed:
mgard: Add configuration for compiling mgard with hip

robertu94 commented:
Rather than submitting this as a patch, please send a PR to MGARD with the patch, and I can merge it early (assuming it works).

kmorel commented:
@robertu94 https://github.com/CODARcode/MGARD/pull/238 is merged into the main repo’s master branch. Did you want to patch all of the [tags](https://github.com/robertu94/MGARD/tags) used by this Spack package, or do you want to keep the patch here to apply to these existing older versions? I think the latter is the more typical Spack solution, but I will leave that up to you.

robertu94 commented:
Ok. I can merge and we can add the patch for the older versions. Is this urgent or are you okay with waiting until January 3rd

robertu94 commented:
Here is the envionment that I built with

```yaml
# This is a Spack Environment file.
#
# It describes a set of packages to be installed, along with
# configuration settings.
spack:
  # add package specs to the `specs` list
  packages:
    mgard:
      variants: +rocm amdgpu_target==gfx941
  specs: 
  - matrix: 
    - [mgard@master, mgard@2024, mgard@2023.12.09]
  view: false
  concretizer:
    unify: false
```

I have a local patch that re-spells the version numbers to use a `.` instead of a `-` so that versions sort correctly.  Master correctly points to a version including your fix for hip.  I'm guessing this is a cmake problem.

robertu94 reviewed (changes requested):
see above comments

robertu94 commented:
@kmorel No I just used HIP as shown in the spack environment.  There is an _aditional_ fake error in the CI related to OneAPI not caused by your patch.

kmorel commented:
@robertu94 OK, I see. It looks like MGARD depends on `hipcub`. I didn't catch that because `hipcub` is usually installed with `ROCm`, but not in your case. I've updated this PR to require hipcub. Hopefully that will fix the issue.

kmorel commented:
@robertu94 Ping.

robertu94 commented:
Sorry for the delayed response @kmorel ,  I pulled in your new change locally, and I am still getting hipcub missing include file errors.  I confirmed that spack re-concretized and that added the new hipcub dependency.  My guess is there is a missing cmake `find_package` call.

Have you tried building with a spack provided hip and hipcub?

kmorel commented:
I see. You are probably right. I will probably have to both modify this PR and submit a new PR to the MGARD project. I will look into this.

kmorel commented:
@robertu94 Can you recommend a docker image that I can use to compile with the base ROCm (highlighting the problems you are having)?

robertu94 commented:
@kmorel I do not have a docker image.  I am just using Fedora 41 (without ROCm installed without spack), and then used the following spack environment above.  It took like 3 hours, but it did eventually build everything except MGARD from source.

kmorel committed:
mgard: Add patch to properly configure hipCUB

kmorel commented:
@robertu94 Can you try running through your compiles again? I think the configuration should load the necessary `hipcub` configuration.

Note that I expect `mgard@2023.12.09` to work, but `mgard@master` will probably fail. The patch I added will not be added to `@master`. Rather, it will rely on MGARD PR https://github.com/CODARcode/MGARD/pull/239. However, that PR is not merged yet. I don't know what `mgard@2024` will do. I don't see that version named.

robertu94 commented:
@kmorel I tried your newest patch this morning, but it did not work for me.  I suspect the problem is with the target_include_directories line that you added because no include paths got added to the cmake generated makefiles.  I confirmed that the patch did get applied; it's just not working.

I couldn't follow what you were saying in your comment regarding the library has chaing the compiler away from HIP.  Could you please elaborate?

kmorel commented:
Aaarg! I really wish AMD didn't have a bare-bones ROCm install.

The problem I was having was with `target_link_libraries`. Normally, I should be able to just link with CMake like this:

```cmake
  target_link_libraries(mgard-library PUBLIC hip::hipcub)
```

That adds all the compile and link flags that should be needed to use the library (including header files) as well as trace down any dependencies. The problem I was having was that when I use this CMake command, the MGARD files are being compiled with gcc instead of the amdclang compiler that they need for the HIP extensions. It seems to be changing the language from HIP to C++, but for the life of me I don't know why.

I'm using `target_include_directories` to try to get around this problem. That should just add the command line arguments to point to the include directories, thereby avoiding any side effects that change the language. However, it does not seem to be grabbing the include directories correctly.

robertu94 commented:
Maybe the maintainers of hipcub can help @afzpatel  @renjithravindrankannath  @srekolam ☝🏻

afzpatel commented:
I tried installing using `spack install mgard +rocm amdgpu_target=gfx906` and I ran into this error:
```
/tmp/root/spack-stage/spack-stage-mgard-2023-12-09-kqbmbareyavtmvo5mg2yzw7z6ozrmlt7/spack-src/include/mgard-x/Hierarchy/../RuntimeX/DeviceAdapters/DeviceAdapterOpenmp.h:14:10: fatal error: 'omp.h' file not found
   14 | #include <omp.h>
      |          ^~~~~~~
```
I think you might be missing a `rocm-openmp-extras` dependency when building with `+openmp+rocm`

I then tried to install using `spack install mgard +rocm ~openmp amdgpu_target=gfx906` with `target_link_libraries(mgard-library PUBLIC hip::hipcub)` but I ran into this issue:
`g++: error: language hip not recognized`
I suspect this is because hip is only supposed to be enabled for ${MGARD_X_HIP_SRC} but now it is being added for the entire mgard-library target. If you take a look at the log you'll see the majority of mgard-library files are being compiled with gcc instead of amdclang++, it is only ${MGARD_X_HIP_SRC} files that are being compiled with amdclang++. This is happening with and without `target_link_libraries(mgard-library PUBLIC hip::hipcub)`.

I'm not sure how  to resolve this, perhaps we can add a mgard-x-hip library with only ${MGARD_X_HIP_SRC} and have the  target_link_libraries call only for the new library?

afzpatel commented:
I was able to solve the hipcub.hpp not found error by using `INTERFACE_INCLUDE_DIRECTORIES` instead of `INCLUDE_DIRECTORIES` but I ended up getting this error:
`/dockerx/spack/opt/spack/linux-ubuntu24.04-skylake_avx512/gcc-13.2.0/rocprim-6.3.0-albnn42xqocon3xittafy6dfu3bba3gf/include/rocprim/device/../detail/../type_traits.hpp:372:1: error: no type named 'type' in 'rocprim::invoke_result<mgard_x::AbsMaxOp, int, double>'
  372 | using invoke_result_t = typename invoke_result<F, Args...>::type;
      | ^~~~~
`

kmorel commented:
Thanks for your help, @afzpatel. Your diagnosis about the gcc error makes sense. I assumed it was using the basic compiler for the GPU code, but it makes sense that it is actually using the GPU compiler for basic C++ code.

For the error you got, are you sure all the rocprim header files were also included? Using `target_link_libraries` will follow all dependencies, but using `target_include_directories` will not. All the headers for all dependent libraries also need to be added. I’m pretty sure that `rocprim::invoke_result` is supposed to have a `type` member. It might be that the correct header is not loaded and the proper definition is missing.

robertu94 commented:
@kmorel is the patch ready for me to take a look again?  I see you did a push Friday?

kmorel commented:
Thanks for asking, @robertu94, but no this is not quite ready yet. I was able to replicate and fix the include directories issue, but then ran into another issue. It's the same one reported by @afzpatel about the missing `type` in `invoke_result`. I don't think MGARD is working with HIP 6.3.1: https://github.com/CODARcode/MGARD/issues/241. I don't know if this is an issue with hipcub or MGARD, but it needs to be fixed regardless.

So, yet one more issue to trace down and fix.

robertu94 commented:
Could you please consider marking it as a draft PR in the meantime @kmorel ?

kmorel commented:
Done.

kmorel committed:
mgard: Add patch to fix issue with HIP collective

kmorel commented:
@robertu94 This PR is now really be ready. I was able to replicate the problems that you and @afzpatel raised, and these are now fixed.

Note that the patches are only applied to the existing versions. For master to work, PRs https://github.com/CODARcode/MGARD/pull/239 and https://github.com/CODARcode/MGARD/pull/242 need to be merged into MGARD (and then that code needs to be mirrored in your https://github.com/robertu94/MGARD repository).

kmorel commented:
@robertu94 Ping.

robertu94 commented:
@kmorel 

I'm still getting errors like this:

```console
==> Error: ProcessError: Command exited with status 2:
    '/home/runderwood/git/contrib/spack/opt/spack/linux-centos7-x86_64_v3/gcc-10.5.0/gmake-4.4.1-x6qzymantv32bmeqgnq25e455a73uoo2/bin/make' '-j10'

8 errors found in build log:
     501    /home/runderwood/git/contrib/spack/opt/spack/linux-fedora41-icelake/gcc-14.2.1/llvm-amdgpu-6.3.2-m4jr6kktrwjhwvpvfokyptzna3tzn5vo/bin/clang++  -DMGARD_ENABLE_HIP -DMGARD_ENABLE_SERIAL -DMGARD_ZSTD -D__HIP_ROCclr__=1 -Dmgard_library_EXPORTS -I
            /home/runderwood/scratch/spackstage/spack-stage-mgard-2023.12.09-vxj4uovaq32tjs7lfngtnis23dxcplw4/spack-src/include -I/home/runderwood/scratch/spackstage/spack-stage-mgard-2023.12.09-vxj4uovaq32tjs7lfngtnis23dxcplw4/spack-build-vxj4uov/includ
            e -isystem /home/runderwood/git/contrib/spack/opt/spack/linux-fedora41-icelake/gcc-14.2.1/protobuf-3.28.2-xypbryobuapjxa5tgsrzcjl2ec7rha6o/include -isystem /home/runderwood/git/contrib/spack/opt/spack/linux-fedora41-icelake/gcc-13.3.1/abseil-
            cpp-20240722.0-dl4pzullwqsjdqzv33b7wvbkyztzoj4d/include -isystem /home/runderwood/git/contrib/spack/opt/spack/linux-centos7-x86_64_v3/gcc-10.5.0/zlib-ng-2.2.3-xxs5bzhp55zr74dym2vw7y7ddreou5c4/include -isystem /home/runderwood/git/contrib/spac
            k/opt/spack/linux-centos7-x86_64_v3/gcc-10.5.0/zstd-1.5.6-7ms2nk43rtoydhzsp427plqalgmd2ze2/include -w -O3 -DNDEBUG -std=gnu++17 --offload-arch=gfx941 -fPIC -DPROTOBUF_USE_DLLS -Wnon-virtual-dtor -DNOMINMAX -o CMakeFiles/mgard-library.dir/src/
            mgard-x/Hierarchy/Hierarchy_1D_double_HIP.cpp.o -x hip -c /home/runderwood/scratch/spackstage/spack-stage-mgard-2023.12.09-vxj4uovaq32tjs7lfngtnis23dxcplw4/spack-build-vxj4uov/src/mgard-x/Hierarchy/Hierarchy_1D_double_HIP.cpp
     502    [ 51%] Building HIP object CMakeFiles/mgard-library.dir/src/mgard-x/Hierarchy/Hierarchy_2D_double_HIP.cpp.o
     503    /home/runderwood/git/contrib/spack/opt/spack/linux-fedora41-icelake/gcc-14.2.1/llvm-amdgpu-6.3.2-m4jr6kktrwjhwvpvfokyptzna3tzn5vo/bin/clang++  -DMGARD_ENABLE_HIP -DMGARD_ENABLE_SERIAL -DMGARD_ZSTD -D__HIP_ROCclr__=1 -Dmgard_library_EXPORTS -I
            /home/runderwood/scratch/spackstage/spack-stage-mgard-2023.12.09-vxj4uovaq32tjs7lfngtnis23dxcplw4/spack-src/include -I/home/runderwood/scratch/spackstage/spack-stage-mgard-2023.12.09-vxj4uovaq32tjs7lfngtnis23dxcplw4/spack-build-vxj4uov/includ
            e -isystem /home/runderwood/git/contrib/spack/opt/spack/linux-fedora41-icelake/gcc-14.2.1/protobuf-3.28.2-xypbryobuapjxa5tgsrzcjl2ec7rha6o/include -isystem /home/runderwood/git/contrib/spack/opt/spack/linux-fedora41-icelake/gcc-13.3.1/abseil-
            cpp-20240722.0-dl4pzullwqsjdqzv33b7wvbkyztzoj4d/include -isystem /home/runderwood/git/contrib/spack/opt/spack/linux-centos7-x86_64_v3/gcc-10.5.0/zlib-ng-2.2.3-xxs5bzhp55zr74dym2vw7y7ddreou5c4/include -isystem /home/runderwood/git/contrib/spac
            k/opt/spack/linux-centos7-x86_64_v3/gcc-10.5.0/zstd-1.5.6-7ms2nk43rtoydhzsp427plqalgmd2ze2/include -w -O3 -DNDEBUG -std=gnu++17 --offload-arch=gfx941 -fPIC -DPROTOBUF_USE_DLLS -Wnon-virtual-dtor -DNOMINMAX -o CMakeFiles/mgard-library.dir/src/
            mgard-x/Hierarchy/Hierarchy_2D_double_HIP.cpp.o -x hip -c /home/runderwood/scratch/spackstage/spack-stage-mgard-2023.12.09-vxj4uovaq32tjs7lfngtnis23dxcplw4/spack-build-vxj4uov/src/mgard-x/Hierarchy/Hierarchy_2D_double_HIP.cpp
     504    In file included from /home/runderwood/scratch/spackstage/spack-stage-mgard-2023.12.09-vxj4uovaq32tjs7lfngtnis23dxcplw4/spack-build-vxj4uov/src/mgard-x/Hierarchy/Hierarchy_1D_double_HIP.cpp:8:
     505    In file included from /home/runderwood/scratch/spackstage/spack-stage-mgard-2023.12.09-vxj4uovaq32tjs7lfngtnis23dxcplw4/spack-src/include/mgard-x/Hierarchy/Hierarchy.hpp:8:
     506    In file included from /home/runderwood/scratch/spackstage/spack-stage-mgard-2023.12.09-vxj4uovaq32tjs7lfngtnis23dxcplw4/spack-src/include/mgard-x/Hierarchy/../RuntimeX/RuntimeX.h:37:
  >> 507    /home/runderwood/scratch/spackstage/spack-stage-mgard-2023.12.09-vxj4uovaq32tjs7lfngtnis23dxcplw4/spack-src/include/mgard-x/Hierarchy/../RuntimeX/DeviceAdapters/DeviceAdapterHip.h:11:10: fatal error: 'hipcub/hipcub.hpp' file not found
     508       11 | #include <hipcub/hipcub.hpp>
     509          |          ^~~~~~~~~~~~~~~~~~~
     510    [ 52%] Building HIP object CMakeFiles/mgard-library.dir/src/mgard-x/Hierarchy/Hierarchy_3D_double_HIP.cpp.o
     511    /home/runderwood/git/contrib/spack/opt/spack/linux-fedora41-icelake/gcc-14.2.1/llvm-amdgpu-6.3.2-m4jr6kktrwjhwvpvfokyptzna3tzn5vo/bin/clang++  -DMGARD_ENABLE_HIP -DMGARD_ENABLE_SERIAL -DMGARD_ZSTD -D__HIP_ROCclr__=1 -Dmgard_library_EXPORTS -I
            /home/runderwood/scratch/spackstage/spack-stage-mgard-2023.12.09-vxj4uovaq32tjs7lfngtnis23dxcplw4/spack-src/include -I/home/runderwood/scratch/spackstage/spack-stage-mgard-2023.12.09-vxj4uovaq32tjs7lfngtnis23dxcplw4/spack-build-vxj4uov/includ
            e -isystem /home/runderwood/git/contrib/spack/opt/spack/linux-fedora41-icelake/gcc-14.2.1/protobuf-3.28.2-xypbryobuapjxa5tgsrzcjl2ec7rha6o/include -isystem /home/runderwood/git/contrib/spack/opt/spack/linux-fedora41-icelake/gcc-13.3.1/abseil-
            cpp-20240722.0-dl4pzullwqsjdqzv33b7wvbkyztzoj4d/include -isystem /home/runderwood/git/contrib/spack/opt/spack/linux-centos7-x86_64_v3/gcc-10.5.0/zlib-ng-2.2.3-xxs5bzhp55zr74dym2vw7y7ddreou5c4/include -isystem /home/runderwood/git/contrib/spac
            k/opt/spack/linux-centos7-x86_64_v3/gcc-10.5.0/zstd-1.5.6-7ms2nk43rtoydhzsp427plqalgmd2ze2/include -w -O3 -DNDEBUG -std=gnu++17 --offload-arch=gfx941 -fPIC -DPROTOBUF_USE_DLLS -Wnon-virtual-dtor -DNOMINMAX -o CMakeFiles/mgard-library.dir/src/
            mgard-x/Hierarchy/Hierarchy_3D_double_HIP.cpp.o -x hip -c /home/runderwood/scratch/spackstage/spack-stage-mgard-2023.12.09-vxj4uovaq32tjs7lfngtnis23dxcplw4/spack-build-vxj4uov/src/mgard-x/Hierarchy/Hierarchy_3D_double_HIP.cpp
     512    1 error generated when compiling for gfx941.
  >> 513    make[2]: *** [CMakeFiles/mgard-library.dir/build.make:2962: CMakeFiles/mgard-library.dir/src/mgard-x/Hierarchy/Hierarchy_1D_double_HIP.cpp.o] Error 1
     514    make[2]: *** Waiting for unfinished jobs....
     515    In file included from /home/runderwood/scratch/spackstage/spack-stage-mgard-2023.12.09-vxj4uovaq32tjs7lfngtnis23dxcplw4/spack-build-vxj4uov/src/mgard-x/Hierarchy/Hierarchy_2D_double_HIP.cpp:8:
     516    In file included from /home/runderwood/scratch/spackstage/spack-stage-mgard-2023.12.09-vxj4uovaq32tjs7lfngtnis23dxcplw4/spack-src/include/mgard-x/Hierarchy/Hierarchy.hpp:8:
     517    In file included from /home/runderwood/scratch/spackstage/spack-stage-mgard-2023.12.09-vxj4uovaq32tjs7lfngtnis23dxcplw4/spack-src/include/mgard-x/Hierarchy/../RuntimeX/RuntimeX.h:37:
  >> 518    /home/runderwood/scratch/spackstage/spack-stage-mgard-2023.12.09-vxj4uovaq32tjs7lfngtnis23dxcplw4/spack-src/include/mgard-x/Hierarchy/../RuntimeX/DeviceAdapters/DeviceAdapterHip.h:11:10: fatal error: 'hipcub/hipcub.hpp' file not found
     519       11 | #include <hipcub/hipcub.hpp>
     520          |          ^~~~~~~~~~~~~~~~~~~
     521    1 error generated when compiling for gfx941.
  >> 522    make[2]: *** [CMakeFiles/mgard-library.dir/build.make:2975: CMakeFiles/mgard-library.dir/src/mgard-x/Hierarchy/Hierarchy_2D_double_HIP.cpp.o] Error 1
     523    In file included from /home/runderwood/scratch/spackstage/spack-stage-mgard-2023.12.09-vxj4uovaq32tjs7lfngtnis23dxcplw4/spack-build-vxj4uov/src/mgard-x/Hierarchy/Hierarchy_3D_double_HIP.cpp:8:
     524    In file included from /home/runderwood/scratch/spackstage/spack-stage-mgard-2023.12.09-vxj4uovaq32tjs7lfngtnis23dxcplw4/spack-src/include/mgard-x/Hierarchy/Hierarchy.hpp:8:
     525    In file included from /home/runderwood/scratch/spackstage/spack-stage-mgard-2023.12.09-vxj4uovaq32tjs7lfngtnis23dxcplw4/spack-src/include/mgard-x/Hierarchy/../RuntimeX/RuntimeX.h:37:
  >> 526    /home/runderwood/scratch/spackstage/spack-stage-mgard-2023.12.09-vxj4uovaq32tjs7lfngtnis23dxcplw4/spack-src/include/mgard-x/Hierarchy/../RuntimeX/DeviceAdapters/DeviceAdapterHip.h:11:10: fatal error: 'hipcub/hipcub.hpp' file not found
     527       11 | #include <hipcub/hipcub.hpp>
     528          |          ^~~~~~~~~~~~~~~~~~~
     529    1 error generated when compiling for gfx941.
  >> 530    make[2]: *** [CMakeFiles/mgard-library.dir/build.make:2988: CMakeFiles/mgard-library.dir/src/mgard-x/Hierarchy/Hierarchy_3D_double_HIP.cpp.o] Error 1
     531    make[2]: Leaving directory '/home/runderwood/scratch/spackstage/spack-stage-mgard-2023.12.09-vxj4uovaq32tjs7lfngtnis23dxcplw4/spack-build-vxj4uov'
  >> 532    make[1]: *** [CMakeFiles/Makefile2:1250: CMakeFiles/mgard-library.dir/all] Error 2
     533    make[1]: Leaving directory '/home/runderwood/scratch/spackstage/spack-stage-mgard-2023.12.09-vxj4uovaq32tjs7lfngtnis23dxcplw4/spack-build-vxj4uov'
  >> 534    make: *** [Makefile:139: all] Error 2

See build log for details:
  /home/runderwood/scratch/spackstage/spack-stage-mgard-2023.12.09-vxj4uovaq32tjs7lfngtnis23dxcplw4/spack-build-out.txt
```

I've confirmed that your patch was correctly applied to CMake.

kmorel commented:
I don't understand how that can be.

Can you share the diff of the patched `CMakeLists.txt`? That is, `diff CMakeLists.txt.orig CMakeLists.txt`.

Can you verify that the `hipcub` package got installed before `mgard` was attempted to be built?

afzpatel commented on var/spack/repos/spack_repo/builtin/packages/mgard/package.py:
```diff
@@ -4,11 +4,12 @@
 
 from spack_repo.builtin.build_systems.cmake import CMakePackage
```
```suggestion
from spack_repo.builtin.build_systems.cmake import CMakePackage, CMakeBuilder
```
I'm getting this error when I tested this pr: `Error: NameError: name 'CMakeBuilder' is not defined`

kwryankrattiger replied:
Yeah the auto-convert missed it, I ran into this as well in my rebase.

kmorel replied:
I replicated this issue as well. Things should work now. I was able to install in a clean Docker container that compiles the hipcub package.

kwryankrattiger commented:
@robertu94 I am not able to reproduce your error with hipcub.

I am building on Frontier which uses a lot of configs, but the things to note are
- gcc@12.2.0
- external HIP

I do get a linking error, but that is likely more because of my configs than it is about the recipe.

afzpatel commented:
> @kwryankrattiger I've only ever built with spack built HIP. My understanding is that there is a bug somewhere in the search directorties for the cmake. @kmorel has been using external HIP which is why I think he's had fewer problems.

I'm also using spack built HIP, and I no longer encounter issues with the hipCUB header files. However, building with `+rocm +openmp` flags is currently broken. Here's the error I'm facing:
```
/tmp/root/spack-stage/spack-stage-mgard-2023-12-09-635qthyosiadlabbjwb4ipwzmjbe64uz/spack-src/include/mgard-x/Hierarchy/../RuntimeX/DeviceAdapters/DeviceAdapterOpenmp.h:14:10: fatal error: 'omp.h' file not found
   14 | #include <omp.h>
      |          ^~~~~~~

```
I was able to build successfully using `+rocm ~openmp`

afzpatel commented:
> > I'm also using spack built HIP, and I no longer encounter issues with the hipCUB header files. However, building with `+rocm +openmp` args is currently broken.
> 
> I've definitely had my issues with compilers not supporting OpenMP, and using `~openmp` is the straightforward way to resolve the issue. If you like, I can make `+rocm` incompatible with `+openmp`. However, it is not clear if the issue is with the mgard package or the configuration of the system. Before I make that change, I would prefer that you try compiling with `~rocm +openmp` to verify that works.

I just tried compiling with `~rocm +openmp`, and it was successful. I agree that making `+rocm` incompatible with `+openmp` would be a good idea, or at least consider setting the default to `~openmp` when `+rocm` is selected.

kmorel committed:
mgard: make +rocm incompatible with +openmp

Currently the OpenMP code is not compiling correctly with the ROCm
compiler. Rather than chase down this problem, I'm just making these
conflict with each other.

kmorel commented:
> Is this ready for review again?

Yes, I believe so.

afzpatel reviewed (approved):
LGTM, just to need to fix the remaining style error

kmorel commented:
> LGTM, just to need to fix the remaining style error

Thanks for pointing that out. The style error has been fixed.

robertu94 reviewed (approved):
We'll need to migrate this to the new spack-packages, repo, but this finally builds cleanly for me.  Thanks for all the hard work @kmorel

kmorel commented:
> We'll need to migrate this to the new spack-packages, repo, but this finally builds cleanly for me. Thanks for all the hard work @kmorel

Nuts. Missed the deadline by one day. I've opened a new PR in the new repo and am closing this one.

@robertu94 Please review https://github.com/spack/spack-packages/pull/44. It is just a copy of this package (except I deleted the declaration of version `master` per the discussion on #50799).
